### PR TITLE
create enumerable module - Pull Request

### PR DESCRIPTION
**Pull request by icebox827**
- Use git-flow to create branches
- create Ruby file
-  Set up linters

### CODE REVIEW

**Review by PhillipUg — COMMENTED** [review]
Hey team, 👋

You have done an amazing job implementing the requirements for this project.

However, there are still some changes 👇 you need to implement to approve this request 🔍

### Changes Required ♻️ 

- [ ] Most of your methods are failing. So i am going to provide you with some sample test cases. Make sure you're methods pass before submitting for review again.
```
puts multiply_els([2, 4, 5]) # => 40
puts

# 1. my_each
puts 'my_each'
puts '-------'
puts [1, 2, 3].my_each { |elem| print "#{elem + 1} " } # => 2 3 4
puts

# 2. my_each_with_index
puts 'my_each_with_index'
puts '------------------'
print [1, 2, 3].my_each_with_index { |elem, idx| puts "#{elem} : #{idx}" } # => 1 : 0, 2 : 1, 3 : 2
puts

# 3. my_select
puts 'my_select'
puts '---------'
p [1, 2, 3, 8].my_select(&:even?) # => [2, 8]
p [0, 2018, 1994, -7].my_select { |n| n > 0 } # => [2018, 1994]
p [6, 11, 13].my_select(&:odd?) # => [11, 13]
puts

# 4. my_all? (example test cases)
puts 'my_all?'
puts '-------'
p [3, 5, 7, 11].my_all?(&:odd?) # => true
p [-8, -9, -6].my_all? { |n| n < 0 } # => true
p [3, 5, 8, 11].my_all?(&:odd?) # => false
p [-8, -9, -6, 0].my_all? { |n| n < 0 } # => false
# test cases required by tse reviewer
p [1, 2, 3, 4, 5].my_all? # => true
p [1, 2, 3].my_all?(Integer) # => true
p %w[dog door rod blade].my_all?(/d/) # => true
p [1, 1, 1].my_all?(1) # => true
puts

# 5. my_any? (example test cases)
puts 'my_any?'
puts '-------'
p [7, 10, 3, 5].my_any?(&:even?) # => true
p [7, 10, 4, 5].my_any?(&:even?) # => true
p %w[q r s i].my_any? { |char| 'aeiou'.include?(char) } # => true
p [7, 11, 3, 5].my_any?(&:even?) # => false
p %w[q r s t].my_any? { |char| 'aeiou'.include?(char) } # => false
# test cases required by tse reviewer
p [3, 5, 4, 11].my_any? # => true
p [1, nil, false].my_any?(1) # => true
p [1, nil, false].my_any?(Integer) # => true
p %w[dog door rod blade].my_any?(/z/) # => false
p [1, 2, 3].my_any?(1) # => true
puts

# 6. my_none? (example test cases)
puts 'my_none?'
puts '--------'
p [3, 5, 7, 11].my_none?(&:even?) # => true
p %w[sushi pizza burrito].my_none? { |word| word[0] == 'a' } # => true
p [3, 5, 4, 7, 11].my_none?(&:even?) # => false
p %w[asparagus sushi pizza apple burrito].my_none? { |word| word[0] == 'a' } # => false
# test cases required by tse reviewer
p [1, 2, 3].my_none? # => false
p [1, 2, 3].my_none?(String) # => true
p [1, 2, 3, 4, 5].my_none?(2) # => false
p [1, 2, 3].my_none?(4) # => true
puts

# 7. my_count (example test cases)
puts 'my_count'
puts '--------'
p [1, 4, 3, 8].my_count(&:even?) # => 2
p %w[DANIEL JIA KRITI dave].my_count { |s| s == s.upcase } # => 3
p %w[daniel jia kriti dave].my_count { |s| s == s.upcase } # => 0
# test cases required by tse reviewer
p [1, 2, 3].my_count # => 3
p [1, 1, 1, 2, 3].my_count(1) # => 3
puts

# 8. my_map
puts 'my_map'
puts '------'
p [1, 2, 3].my_map { |n| 2 * n } # => [2,4,6]
p %w[Hey Jude].my_map { |word| word + '?' } # => ["Hey?", "Jude?"]
p [false, true].my_map(&:!) # => [true, false]
my_proc = proc { |num| num > 10 }
p [18, 22, 5, 6].my_map(my_proc) { |num| num < 10 } # => true true false false
puts

# 9. my_inject
puts 'my_inject'
puts '---------'
p [1, 2, 3, 4].my_inject(10) { |accum, elem| accum + elem } # => 20
p [1, 2, 3, 4].my_inject { |accum, elem| accum + elem } # => 10
p [5, 1, 2].my_inject('+') # => 8
p (5..10).my_inject(2, :*) # should return 302400
p (5..10).my_inject(4) { |prod, n| prod * n } # should return 604800

```

#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**
I would be happy to help! 👩‍💻 

Happy coding 🤓

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+# rubocop:disable Metrics/PerceivedComplexity
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given
+
+    arr.length.times do |n|
+      yield arr[n]
+    end
+  end
```
Comment (arslanbisharat) [inline enumerable.rb:11]: - [ ] enumerable #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self

Here is the error I am getting

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  def my_each_with_index
+    return to_enum unless block_given
+
+    arr.length.times do |n, i|
+      yield(arr[i], n)
+    end
+  end
```
Comment (arslanbisharat) [inline enumerable.rb:19]: - [ ] enumerable #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self

Here is the error I am getting

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    return to_enum unless block_given
+
+    list = []
+
+    arr.my_each | filter |
+      if filter != 'accepted'
+        list.push(filter)
+        yield filter
+      end
+  end
```
Comment (arslanbisharat) [inline enumerable.rb:31]: - [ ] enumerable #my_select has three errors. Please have a look at the attachment and remove those errors.

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    speech = false
+
+    if !block_given?
+      my_each { |argument| speech = true unless argument }
+    elsif !args[0].nil?
+      my_each { |argument| speech = true unless args[0] == argument }
+    else
+      my_each { |argument| speech = true unless yield argument }
+    end
+  end
```
Comment (arslanbisharat) [inline enumerable.rb:59]: - [ ] Here are the following errors

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    speech = true
+
+    if !block_given?
+      my_each { |argument| speech = false unless argument }
+    elsif !args[0].nil?
+      my_each { |argument| speech = false unless args[0] == argument }
+    else
+      my_each { |argument| speech = false unless yield argument }
+    end
+  end
```
Comment (arslanbisharat) [inline enumerable.rb:45]: - [ ] There are a number of tests falling in my_all. Please have a look

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+  def my_count(num)
+    return to_enum unless block_given
+
+    count = 0
+
+    my_each { |i| count += 1 if num == i }
+    count
+  end
```
Comment (arslanbisharat) [inline enumerable.rb:82]: - [ ] My_count has following errors

**Review by arslanbisharat — COMMENTED** [review]
# Request changes 😳

I am very happy to see this project and I can see you worked hard on this project. There are only a few issues that we need to address and I am sure that you will end up with a perfect project

_It's okay. We are here to learn. Don't get discouraged. Work even harder. You are almost there.😄_

This is a great opportunity to learn. Microverse is here to help you. 

**Please have a look at the suggestions to get more understanding.**

- [ ] You have 41 tests failing. As mentioned by the previous TSE, you should use the file the respected TSE shared with you and then you run those tests against your code.  After all the tests get the pass, please submit another review. 


**Please make the changes and request a re-review for the given project**

If you get stuck somewhere, don't hesitate to reach me out or schedule s 1:1 meeting with a Full-time TSE. 

Stay safe and give me a smile 😇

### Happy Coding
---------------------------------------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

**Pull request merged**